### PR TITLE
Upgrade PromotionCoupon resource to APIP3

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A
|-----------------|-----
| Branch?         | api-platform-3
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 1]
```diff
@@ -328,22 +328,22 @@
...
+            </argument>
+            <tag name="api_platform.filter" />
+        </service>
 
-<!--        <service id="Sylius\Bundle\ApiBundle\Filter\Doctrine\PromotionCouponPromotionFilter" public="true">-->
-<!--            <argument type="service" id="api_platform.symfony.iri_converter" />-->
-<!--            <argument type="service" id="doctrine" />-->
-<!--            <tag name="api_platform.filter" />-->
-<!--        </service>-->
+        <service id="Sylius\Bundle\ApiBundle\Filter\Doctrine\PromotionCouponPromotionFilter" public="true">
```
Comment (mpysiak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml]:
```suggestion
        <service id="sylius_api.filter.promotion_coupon.promotion" class="Sylius\Bundle\ApiBundle\Filter\Doctrine\PromotionCouponPromotionFilter" public="true">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/PromotionCoupon.xml [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                    </values>
+                </value>
+            </values>
+        </validationContext>
+
+        <operations>
+            <operation class="ApiPlatform\Metadata\GetCollection" uriTemplate="/admin/promotion-coupons">
+                <filters>
+                    <filter>sylius.api.promotion_coupon_order_filter</filter>
+                    <filter>Sylius\Bundle\ApiBundle\Filter\Doctrine\PromotionCouponPromotionFilter</filter>
```
Comment (mpysiak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/PromotionCoupon.xml]:
```suggestion
                    <filter>sylius_api.filter.promotion_coupon.promotion</filter>
```

File: src/Sylius/Bundle/ApiBundle/Serializer/Normalizer/GeneratedPromotionCouponsNormalizer.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    public function __construct(private SectionProviderInterface $sectionProvider)
+    {
+    }
+
+    public function normalize($object, $format = null, array $context = [])
+    {
+        Assert::isInstanceOf($object, ArrayCollection::class);
+        Assert::isInstanceOf($object->first(), PromotionCouponInterface::class);
+        Assert::keyNotExists($context, self::ALREADY_CALLED);
+        Assert::isInstanceOf($this->sectionProvider->getSection(), AdminApiSection::class);
```
Comment (mpysiak) [inline src/Sylius/Bundle/ApiBundle/Serializer/Normalizer/GeneratedPromotionCouponsNormalizer.php:39]:
```suggestion
```

**Review by mpysiak — APPROVED** [review]

**Review by mpysiak — CHANGES_REQUESTED** [review]
Some 500 errors in GA

File: src/Sylius/Bundle/ApiBundle/Serializer/Normalizer/GeneratedPromotionCouponsNormalizer.php (hunk 3 above)
Comment (TheMilek) [reply src/Sylius/Bundle/ApiBundle/Serializer/Normalizer/GeneratedPromotionCouponsNormalizer.php:39]: Why?

Comment (mpysiak) [reply src/Sylius/Bundle/ApiBundle/Serializer/Normalizer/GeneratedPromotionCouponsNormalizer.php:39]: This was already checked is `supportsNormalization` method. And `sectionProvider` it is not used in normalize method

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Promotion.xml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        <validationContext>
+            <values>
+                <value name="groups">
+                    <values>
+                        <value>sylius</value>
+                        <value>sylius_api</value>
+                    </values>
+                </value>
+            </values>
+        </validationContext>
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Promotion.xml]: Why do you need this?

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 5]
```diff
@@ -328,22 +328,22 @@
...
-<!--            <argument type="collection">-->
-<!--                <argument key="code" />-->
-<!--                <argument key="expiresAt" />-->
-<!--                <argument key="usageLimit" />-->
-<!--                <argument key="perCustomerUsageLimit" />-->
-<!--                <argument key="used" />-->
-<!--            </argument>-->
-<!--            <tag name="api_platform.filter" />-->
-<!--        </service>-->
+        <service id="sylius.api.promotion_coupon_order_filter" parent="api_platform.doctrine.orm.order_filter" public="true">
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml]:
```suggestion
        <service id="sylius.api.order_filter.promotion_coupon" parent="api_platform.doctrine.orm.order_filter" public="true">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 6]
```diff
@@ -328,22 +328,22 @@
...
+            </argument>
+            <tag name="api_platform.filter" />
+        </service>
 
-<!--        <service id="Sylius\Bundle\ApiBundle\Filter\Doctrine\PromotionCouponPromotionFilter" public="true">-->
-<!--            <argument type="service" id="api_platform.symfony.iri_converter" />-->
-<!--            <argument type="service" id="doctrine" />-->
-<!--            <tag name="api_platform.filter" />-->
-<!--        </service>-->
+        <service id="sylius_api.filter.promotion_coupon.promotion" class="Sylius\Bundle\ApiBundle\Filter\Doctrine\PromotionCouponPromotionFilter" public="true">
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml]:
```suggestion
        <service id="sylius_api.doctrine.orm.filter.promotion_coupon_promotion" class="Sylius\Bundle\ApiBundle\Doctrine\ORM\Filters\PromotionCouponPromotionFilter" public="true">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/serializers.xml [hunk 7]
```diff
@@ -100,6 +100,11 @@
             <tag name="serializer.normalizer" priority="64" />
         </service>
 
+        <service id="sylius_api.normalizer.promotion_coupon" class="Sylius\Bundle\ApiBundle\Serializer\Normalizer\GeneratedPromotionCouponsNormalizer">
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/serializers.xml]:
```suggestion
        <service id="sylius_api.normalizer.generated_promotion_coupons" class="Sylius\Bundle\ApiBundle\Serializer\Normalizer\GeneratedPromotionCouponsNormalizer">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/PromotionCoupon.xml [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+-->
+
+<resources
+    xmlns="https://api-platform.com/schema/metadata/resources-3.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="https://api-platform.com/schema/metadata/resources-3.0 https://api-platform.com/schema/metadata/resources-3.0.xsd"
+>
+
+    <!-- Admin Section -->
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/PromotionCoupon.xml]: Do you have shop section?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/PromotionCoupon.xml [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    <resource class="Sylius\Component\Core\Model\PromotionCoupon">
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/PromotionCoupon.xml]: IMO, the subresource section should come after the resource section

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/PromotionCoupon.xml [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    <resource class="Sylius\Component\Core\Model\PromotionCoupon" shortName="PromotionCoupon">
+        <validationContext>
+            <values>
+                <value name="groups">
+                    <values>
+                        <value>sylius</value>
+                    </values>
+                </value>
+            </values>
+        </validationContext>
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/PromotionCoupon.xml]:
```suggestion
```
Add separate validation groups to each `Post` and `Put` operation.

**Review by Wojdylak — APPROVED** [review]

**Review by mpysiak — APPROVED** [review]

**Pull request merged**